drone/src/proxy/service.rs

```diff
@@ -256,84 +256,102 @@ impl ProxyService {
+    }
 
-                    return Ok(Response::builder()
-                        .status(StatusCode::BAD_REQUEST)
-                        .body("Missing token.".into())?);
-                }
+    fn handle_plane_auth(&self, req: Request<Body>) -> anyhow::Result<Response<Body>> {
+        let params: PlaneAuthParams =
+            serde_html_form::from_str(req.uri().query().unwrap_or_default())?;
 
-                if let Some(proxy_route) = route {
-                    self.connection_tracker.track_request(&subdomain);
-                    *req.uri_mut() = Self::rewrite_uri(&proxy_route.address, req.uri())?;
+        if let Some(redirect) = params.redirect.as_deref() {
+            if !redirect.starts_with('/') {
+                return Ok(Response::builder()
+                    .status(StatusCode::BAD_REQUEST)
+                    .body("Redirect must be relative and start with a slash.".into())
+                    .unwrap());
+            }
+        }
 
-                    if let Some(token) = proxy_route.bearer_token {
-                        if let Some(response) = check_auth(&req, &token)? {
-                            return Ok(response);
-                        }
-                    }
+        Ok(Response::builder()
+            .status(StatusCode::FOUND)
+            .header("Location", params.redirect.as_deref().unwrap_or("/"))
+            .header(
+                "Set-Cookie",
+                format!("{}={}", PLANE_AUTH_COOKIE, params.token),
+            )
+            .body(Body::empty())?)
+    }
 
-                    if let Some(connection) = req.headers().get(hyper::http::header::CONNECTION) {
-                        if connection
-                            .to_str()
-                            .unwrap_or_default()
-                            .to_lowercase()
-                            .contains(UPGRADE)
-                        {
-                            return self.handle_upgrade(req, &subdomain).await;
-                        }
-                    }
+    async fn handle(self, mut req: Request<Body>) -> anyhow::Result<Response<Body>> {
+        let subdomain = match self.backend_from_request(&req) {
+            Ok(subdomain) => subdomain,
+            Err(error) => {
+                tracing::error!(?error, "Error getting backend from request.");
+                return Ok(Response::builder()
+                    .status(StatusCode::BAD_REQUEST)
+                    .header(hyper::header::ACCESS_CONTROL_ALLOW_ORIGIN, "*")
+                    .body(Body::empty())?);
+            }
+        };
+
+        tracing::info!(ip=%self.remote_ip, url=%req.uri(), "Proxy Request");
+        let route = self.db.get_proxy_route(&subdomain).await?.or_else(|| {
+            self.passthrough.map(|d| ProxyRoute {
+                address: d.to_string(),
+                bearer_token: None,
+            })
+        });
+
+        if req.uri().path() == "/_plane_auth" {
```

Comment:
Will this work with the `/_plane_backend=` prefix because `path` was mutated in `backend_from_request`? (if so, does `req.uri().path()` return that same mutated path?) And would this code break if the `handle_plane_auth()` step was moved to the top of this function?

Maybe `backend_from_request` should return the backend _and_ the stripped path?

Reply:
Good catch, I've made `backend_from_request` modify the URL in place. I don't love it as a side-effect, but since it's just for a dev feature I'm ok with it. I'd like to refactor the whole thing at some point.